packages/babel-parser/src/plugins/typescript/index.js

```diff
@@ -2952,9 +2955,36 @@ export default (superClass: Class<Parser>): Class<Parser> =>
           this.resetStartLocationFromNode(expr, typeParameters);
         }
         expr.typeParameters = typeParameters;
+
+        // report error if single type parameter used without trailing comma.
+        if (
+          this.hasPlugin("jsx") &&
+          expr.typeParameters.params.length === 1 &&
+          !expr.typeParameters.extra?.trailingComma
+        ) {
+          const parameter = expr.typeParameters.params[0];
+          if (parameter.constraint) {
+            // If parameter has any constraints, it must contain multiple tokens.
+            // <T extends U> is a valid declaration.
+            // <T extends {name: string}> is also a valid declaration.
+          } else invalidSingleType = parameter;
```

Comment:
Nit:
```suggestion
          if (!parameter.constraint) {
            // A single type parameter must either have constraints
            // or a trailing comma, otherwise it's ambiguous with JSX.
            invalidSingleType = parameter;
          }
```